source/PlayerInfoPanel.cpp

```diff
@@ -654,13 +796,78 @@ void PlayerInfoPanel::DrawFleet(const Rectangle &bounds)
 
 
 
+// Sorts the player's fleet given a comparator function (based on column)
+void PlayerInfoPanel::SortShips(InfoPanelState::ShipComparator &shipComparator)
+{
+	// Save selected ships to preserve selection after sort
+	set<shared_ptr<Ship>> selectedShips;
+	shared_ptr<Ship> lastSelected = panelState.SelectedIndex() == -1
+	? nullptr
+	: panelState.Ships()[panelState.SelectedIndex()];
+	
+	for(int i : panelState.AllSelected())
+		selectedShips.insert(panelState.Ships()[i]);
+	panelState.AllSelected().clear();
+	
+	// Move flagship to first position
+	for(auto it = panelState.Ships().begin(); it != panelState.Ships().end(); ++it)
+		if(it->get() == player.Flagship())
+		{
+			iter_swap(it, panelState.Ships().begin());
+			break;
+		}
+	
+	// If ships are not sorted according to this comparator,
+	// then sort them, otherwise reverse the sort order.
+	if(panelState.CurrentSort() != shipComparator)
+		std::stable_sort(
+			panelState.Ships().begin() + 1,
+			panelState.Ships().end(),
+			[&](const shared_ptr<Ship> &lhs, const  shared_ptr<Ship> &rhs
+		){
+			return shipComparator(lhs, rhs);
+		});
+	else
+		std::reverse(panelState.Ships().begin() + 1, panelState.Ships().end());
+	
+	// Load the same selected ships from before the sort
+	for(unsigned int i = 0; i < panelState.Ships().size(); i++)
+	{
+		auto ship = panelState.Ships()[i];
+		if(selectedShips.count(ship))
```

Comment:
For large lists of ships and/or selections, it will be more performant to use `find` and then erase the particular iterator position, rather than use both `count(key)` and `erase(key)`.